###### Enable saving of the notebook every 5 minutes

In [1]:
%autosave 5

Autosaving every 5 seconds


###### Import necessary modules

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib as mpl
import matplotlib.pyplot as plot
%matplotlib inline

In [3]:
# Read Excel file
xl_file = pd.ExcelFile('Data analytics project dataset.xlsx')

In [4]:
sheet_names = xl_file.sheet_names # Get sheet names of the Excel file

In [5]:
# Extract all sheets as dataframes in the file into a dictionary 
# With sheet name as key and dataframe as value
dfs = {sheet_name: xl_file.parse(sheet_name) 
          for sheet_name in sheet_names}

In [6]:
dfs['Data Science Hackathon dataset'] # Get a sheet name by name

InsurProductID|InsurCustomerID|Sub Seg|Insurance Currency|Insur Sub Name|Insur Bank Related|Insur PackName|Product Status|Insur Value|Insur Open Date|Insur Cancellation Date|Insur end date|Next Pay Date|Last Pay Date|NO_OF_TRXNS202005|NO_OF_TRXNS202006|NO_OF_TRXNS202007|NO_OF_TRXNS202008|NO_OF_TRXNS202009|NO_OF_TRXNS202010|TURNOVER202005|TURNOVER202006|TURNOVER202007|TURNOVER202008|TURNOVER202009|TURNOVER202010  \
0       LFC484L84|475143751|EXCELLENCE|MZN|SEGURO DE V...                                                                                                                                                                                                                                                                                                                                                                               
1       HOC484H84|475143751|EXCELLENCE|MZN|SEGURO DE I...                                                                                                                                                                                                                                                                                                                                                                               
2       SBS025S25|628865188|GENNESIS|MZN|PLANO SALARIO...                                                                                                                                                                                                                                                                                                                                                                               
3       SBS149S49|169913599|PMO2|MZN||NON BANK RELATED...                                                                                                                                                                                                                                                                                                                                                                               
4       SBS005S05|558654486|FAMMILY|MZN||NON BANK RELA...                                                                                                                                                                                                                                                                                                                                                                               
...                                                   ...                                                                                                                                                                                                                                                                                                                                                                               
139385  IPN710I10|953294832|FAMMILY|MZN|SEGURO DE CRÉD...                                                                                                                                                                                                                                                                                                                                                                               
139386  IPS049I49|983790337|FAMMILY|MZN|SEGURO DE CRÉD...                                                                                                                                                                                                                                                                                                                                                                               
139387  LFN133L33|905290252|PLATIUM|MZN|SEGURO DE VIDA...                                                                                                                             

As the above table can be seen, the data is not structured. We need to do <b>Data Cleansing</b>

In [7]:
# Check column names in the dataset
dfs['Data Science Hackathon dataset'].columns

Index(['InsurProductID|InsurCustomerID|Sub Seg|Insurance Currency|Insur Sub Name|Insur Bank Related|Insur PackName|Product Status|Insur Value|Insur Open Date|Insur Cancellation Date|Insur end date|Next Pay Date|Last Pay Date|NO_OF_TRXNS202005|NO_OF_TRXNS202006|NO_OF_TRXNS202007|NO_OF_TRXNS202008|NO_OF_TRXNS202009|NO_OF_TRXNS202010|TURNOVER202005|TURNOVER202006|TURNOVER202007|TURNOVER202008|TURNOVER202009|TURNOVER202010', 'Unnamed: 1'], dtype='object')

We see that the dataframe has two columns, but not correctly formatted. We need to extract the first column,
where the actual data exist.

In [8]:
# Get 
df2 = dfs['Data Science Hackathon dataset']['InsurProductID|InsurCustomerID|Sub Seg|Insurance Currency|Insur Sub Name|Insur Bank Related|Insur PackName|Product Status|Insur Value|Insur Open Date|Insur Cancellation Date|Insur end date|Next Pay Date|Last Pay Date|NO_OF_TRXNS202005|NO_OF_TRXNS202006|NO_OF_TRXNS202007|NO_OF_TRXNS202008|NO_OF_TRXNS202009|NO_OF_TRXNS202010|TURNOVER202005|TURNOVER202006|TURNOVER202007|TURNOVER202008|TURNOVER202009|TURNOVER202010']

In [9]:
# Get column names
columns = dfs['Data Science Hackathon dataset'].columns[0]
# Split column names, as they are stored in a single string separated by |.
columns = columns.split('|')

The dataframe, df2, need to be split by separator (|) so that we can have rows correctly structured.

In [10]:
data = [col for col in df2.str.split('|')]

In [11]:
# Create dataframe with the 
df = pd.DataFrame(data, columns=columns)

df.head() # Get the first 5 rows in the dataframe

InsurProductID InsurCustomerID     Sub Seg Insurance Currency  \
0      LFC484L84       475143751  EXCELLENCE                MZN   
1      HOC484H84       475143751  EXCELLENCE                MZN   
2      SBS025S25       628865188    GENNESIS                MZN   
3      SBS149S49       169913599        PMO2                MZN   
4      SBS005S05       558654486     FAMMILY                MZN   

                 Insur Sub Name     Insur Bank Related Insur PackName  \
0                SEGURO DE VIDA  BANK PRODUCTS RELATED           LIFE   
1  SEGURO DE INCÊNDIO HABITAÇÃO  BANK PRODUCTS RELATED    HOUSE.OWNER   
2                 PLANO SALARIO       NON BANK RELATED        SP.BLUE   
3                                     NON BANK RELATED                  
4                                     NON BANK RELATED                  

  Product Status Insur Value        Insur Open Date  ... NO_OF_TRXNS202007  \
0         ACTIVE        0.00  3/26/2013 12:00:00 AM  ...                61   
1         ACTIVE        0.00  3/26/2013 12:00:00 AM  ...                61   
2      CANCELLED        0.00   7/3/2018 12:00:00 AM  ...                 8   
3         ACTIVE        0.00  7/31/2019 12:00:00 AM  ...                     
4         ACTIVE        0.00  9/12/2019 12:00:00 AM  ...                24   

  NO_OF_TRXNS202008 NO_OF_TRXNS202009 NO_OF_TRXNS202010   TURNOVER202005  \
0                83                61                59  238558.20284615   
1                83                61                59  238558.20284615   
2                 7                18                11  7968.0634615385   
3                                                                          
4                31                26                52  370471.99719231   

    TURNOVER202006 TURNOVER202007   TURNOVER202008   TURNOVER202009  \
0  262221.48884615     301020.335   229569.0262963  226266.91481481   
1  262221.48884615     301020.335   229569.0262963  226266.91481481   
2  7377.8365384615         8043.9  9608.7532222222  9207.5236296296   
3                                                                     
4  251616.88884615      61927.235        73461.539        53934.812   

    TURNOVER202010  
0  217867.91177419  
1  217867.91177419  
2  8935.0903225806  
3                   
4  145213.38690323  

[5 rows x 26 columns]

<h4>Evaluating for Missing Data in the Dataframe</h4>

The missing values are converted by default. There are two methods to detect missing data:

<ol>
    <li><b>.isnull()</b></li>
    <li><b>.notnull()</b></li>
</ol>
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.

In [12]:
# Let's get missing values in the dataframe
missing_data = df.isnull()
missing_data.head()

InsurProductID  InsurCustomerID  Sub Seg  Insurance Currency  \
0           False            False    False               False   
1           False            False    False               False   
2           False            False    False               False   
3           False            False    False               False   
4           False            False    False               False   

   Insur Sub Name  Insur Bank Related  Insur PackName  Product Status  \
0           False               False           False           False   
1           False               False           False           False   
2           False               False           False           False   
3           False               False           False           False   
4           False               False           False           False   

   Insur Value  Insur Open Date  ...  NO_OF_TRXNS202007  NO_OF_TRXNS202008  \
0        False            False  ...              False              False   
1        False            False  ...              False              False   
2        False            False  ...              False              False   
3        False            False  ...              False              False   
4        False            False  ...              False              False   

   NO_OF_TRXNS202009  NO_OF_TRXNS202010  TURNOVER202005  TURNOVER202006  \
0              False              False           False           False   
1              False              False           False           False   
2              False              False           False           False   
3              False              False           False           False   
4              False              False           False           False   

   TURNOVER202007  TURNOVER202008  TURNOVER202009  TURNOVER202010  
0           False           False           False           False  
1           False           False           False           False  
2           False           False           False           False  
3           False           False           False           False  
4           False           False           False           False  

[5 rows x 26 columns]

<b>True</b> means the value is a missing value while <b>False</b> means the value is not a missing value.


<h4>Count missing values in each column</h4>
<p>
Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, <b>True</b> represents a missing value and <b>False</b> means the value is present in the dataset.  In the body of the for loop the method <b>.value_counts()</b> counts the number of <b>True</b> values. 
</p>


In [13]:
# Check Percentage missing values for each column

for column in missing_data.columns.values.tolist():
    print(column)
    # Count missing values (True or False) for each column
    print(missing_data[column].value_counts())
    print('')

InsurProductID
False    139390
Name: InsurProductID, dtype: int64

InsurCustomerID
False    139390
Name: InsurCustomerID, dtype: int64

Sub Seg
False    139390
Name: Sub Seg, dtype: int64

Insurance Currency
False    139390
Name: Insurance Currency, dtype: int64

Insur Sub Name
False    139390
Name: Insur Sub Name, dtype: int64

Insur Bank Related
False    139390
Name: Insur Bank Related, dtype: int64

Insur PackName
False    139390
Name: Insur PackName, dtype: int64

Product Status
False    139390
Name: Product Status, dtype: int64

Insur Value
False    139390
Name: Insur Value, dtype: int64

Insur Open Date
False    128614
True      10776
Name: Insur Open Date, dtype: int64

Insur Cancellation Date
False    128614
True      10776
Name: Insur Cancellation Date, dtype: int64

Insur end date
False    128614
True      10776
Name: Insur end date, dtype: int64

Next Pay Date
False    128614
True      10776
Name: Next Pay Date, dtype: int64

Last Pay Date
False    128614
True      10776
Nam

Based on the summary above, each column has 139390 rows of data and 17 of the columns containing missing data:
- <b>Insur Open Date, Insur Cancellation Date, Insur end date, Next Pay Date, Last Pay Date</b>
- All column names with the format <b>NO_OF_TRxxxxx</b>
- All column names with the formas <b>TURNOVER2xxxx</b>

All the above columns with missing data, have each 10776 missing values

<h3 id="deal_missing_values">Dealing with missing data</h3>
<b>How to deal with missing data?</b>

<ol>
    <li>Drop data<br>
        a. Drop the whole row<br>
        b. Drop the whole column
    </li>
    <li>Replace data<br>
        a. Replace it by mean<br>
        b. Replace it by frequency<br>
        c. Replace it based on other functions
    </li>
</ol>


Whole columns should be dropped only if most entries in the column are empty. In our dataset, none of the columns are empty enough to drop entirely.
We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method to many different columns:

<b>Replace by mean:</b>

<ul>
    <li>All columns with missing dates that are datetime, we can convert them to the mean date</li>
    <li>Remaining columns with missing date that are not date, we can replace null values with the mean</li>
</ul>

In [14]:
# Replace all empty string values by null values (np.nan)
df.replace('', np.nan, inplace=True)

In [15]:
def replace_missing(df, columns, date_type):
    for col in columns:
        avg = df[col].astype(date_type).mean(axis=0)
        print(f'Average date of {col}: {avg}')
        df[col].replace(np.nan, avg, inplace=True)

        df[col] = df[col].astype(date_type)

<h4>Calculate the mean value for all dates columns and replace missing values with the mean</h4>

In [16]:
date_columns = [col for col in columns if 
                col.endswith('Date') or col.endswith('date')]

# Handle missing values for date columns
replace_missing(df, date_columns, 'datetime64[ns]')

Average date of Insur Open Date: 2019-02-11 01:19:21.744512512
Average date of Insur Cancellation Date: 2020-05-27 19:08:59.629065216
Average date of Insur end date: 2020-06-17 01:56:38.965873920
Average date of Next Pay Date: 2020-10-03 23:32:05.687347456
Average date of Last Pay Date: 2020-08-04 00:37:03.828845312


In [17]:
df[date_columns].dtypes

Insur Open Date            datetime64[ns]
Insur Cancellation Date    datetime64[ns]
Insur end date             datetime64[ns]
Next Pay Date              datetime64[ns]
Last Pay Date              datetime64[ns]
dtype: object

In [18]:
# for col in date_columns:
#     print(df[df[col].isnull()])

<h4>Calculate the mean value for all numerical columns and replace missing values with the mean</h4>

In [19]:
# # Get all column names starting with NO_OF 
# # (which all have missing values)
numerical_columns = [col for col in columns if col.startswith('NO_OF')]

# Handle missing values for column names starting with
# NO_OF
replace_missing(df, numerical_columns, 'float64')

Average date of NO_OF_TRXNS202005: 32.58719631265063
Average date of NO_OF_TRXNS202006: 31.457919180901783
Average date of NO_OF_TRXNS202007: 29.904394560455994
Average date of NO_OF_TRXNS202008: 32.36226976056859
Average date of NO_OF_TRXNS202009: 35.31844730221839
Average date of NO_OF_TRXNS202010: 36.724820998187994


In [20]:
df[numerical_columns].dtypes

NO_OF_TRXNS202005    float64
NO_OF_TRXNS202006    float64
NO_OF_TRXNS202007    float64
NO_OF_TRXNS202008    float64
NO_OF_TRXNS202009    float64
NO_OF_TRXNS202010    float64
dtype: object

In [21]:
# for col in numerical_columns:
#     print(df[df[col].isnull()])

In [22]:
df[df['NO_OF_TRXNS202008'].isnull()]

Empty DataFrame
Columns: [InsurProductID, InsurCustomerID, Sub Seg, Insurance Currency, Insur Sub Name, Insur Bank Related, Insur PackName, Product Status, Insur Value, Insur Open Date, Insur Cancellation Date, Insur end date, Next Pay Date, Last Pay Date, NO_OF_TRXNS202005, NO_OF_TRXNS202006, NO_OF_TRXNS202007, NO_OF_TRXNS202008, NO_OF_TRXNS202009, NO_OF_TRXNS202010, TURNOVER202005, TURNOVER202006, TURNOVER202007, TURNOVER202008, TURNOVER202009, TURNOVER202010]
Index: []

[0 rows x 26 columns]

In [24]:
numerical_columns2 = [col for col in columns if col.startswith('TURNOVER')]

# Handle missing values for column names starting with
# TURNOVER
replace_missing(df, numerical_columns2, 'float64')

Average date of TURNOVER202005: 643642.929147859
Average date of TURNOVER202006: 2443806.898664492
Average date of TURNOVER202007: 2070891.0775850501
Average date of TURNOVER202008: 429646.6086796057
Average date of TURNOVER202009: 398152.76217707945
Average date of TURNOVER202010: 533068.2916875185


In [25]:
df.dtypes

InsurProductID                     object
InsurCustomerID                    object
Sub Seg                            object
Insurance Currency                 object
Insur Sub Name                     object
Insur Bank Related                 object
Insur PackName                     object
Product Status                     object
Insur Value                        object
Insur Open Date            datetime64[ns]
Insur Cancellation Date    datetime64[ns]
Insur end date             datetime64[ns]
Next Pay Date              datetime64[ns]
Last Pay Date              datetime64[ns]
NO_OF_TRXNS202005                 float64
NO_OF_TRXNS202006                 float64
NO_OF_TRXNS202007                 float64
NO_OF_TRXNS202008                 float64
NO_OF_TRXNS202009                 float64
NO_OF_TRXNS202010                 float64
TURNOVER202005                    float64
TURNOVER202006                    float64
TURNOVER202007                    float64
TURNOVER202008                    

By checking the data type of the columns we see that all columns are of type <b>object</b>. We need to convert data types to correct types. For example, a column with integer values must be of type <b>int</b>.

In [26]:
df.head()

InsurProductID InsurCustomerID     Sub Seg Insurance Currency  \
0      LFC484L84       475143751  EXCELLENCE                MZN   
1      HOC484H84       475143751  EXCELLENCE                MZN   
2      SBS025S25       628865188    GENNESIS                MZN   
3      SBS149S49       169913599        PMO2                MZN   
4      SBS005S05       558654486     FAMMILY                MZN   

                 Insur Sub Name     Insur Bank Related Insur PackName  \
0                SEGURO DE VIDA  BANK PRODUCTS RELATED           LIFE   
1  SEGURO DE INCÊNDIO HABITAÇÃO  BANK PRODUCTS RELATED    HOUSE.OWNER   
2                 PLANO SALARIO       NON BANK RELATED        SP.BLUE   
3                           NaN       NON BANK RELATED            NaN   
4                           NaN       NON BANK RELATED            NaN   

  Product Status Insur Value Insur Open Date  ... NO_OF_TRXNS202007  \
0         ACTIVE        0.00      2013-03-26  ...         61.000000   
1         ACTIVE        0.00      2013-03-26  ...         61.000000   
2      CANCELLED        0.00      2018-07-03  ...          8.000000   
3         ACTIVE        0.00      2019-07-31  ...         29.904395   
4         ACTIVE        0.00      2019-09-12  ...         24.000000   

  NO_OF_TRXNS202008 NO_OF_TRXNS202009 NO_OF_TRXNS202010  TURNOVER202005  \
0          83.00000         61.000000         59.000000   238558.202846   
1          83.00000         61.000000         59.000000   238558.202846   
2           7.00000         18.000000         11.000000     7968.063462   
3          32.36227         35.318447         36.724821   643642.929148   
4          31.00000         26.000000         52.000000   370471.997192   

   TURNOVER202006  TURNOVER202007  TURNOVER202008  TURNOVER202009  \
0    2.622215e+05    3.010203e+05   229569.026296   226266.914815   
1    2.622215e+05    3.010203e+05   229569.026296   226266.914815   
2    7.377837e+03    8.043900e+03     9608.753222     9207.523630   
3    2.443807e+06    2.070891e+06   429646.608680   398152.762177   
4    2.516169e+05    6.192724e+04    73461.539000    53934.812000   

   TURNOVER202010  
0   217867.911774  
1   217867.911774  
2     8935.090323  
3   533068.291688  
4   145213.386903  

[5 rows x 26 columns]